test/e2e/reboot.go

```diff
@@ -63,53 +63,102 @@ var _ = Describe("Reboot", func() {
 		Expect(err).NotTo(HaveOccurred())
 	})
 
-	It("should reboot each node and ensure they function upon restart", func() {
-		// This test requires SSH, so the provider check should be identical to
-		// there (the limiting factor is the implementation of util.go's
-		// getSigner(...)).
-		provider := testContext.Provider
-		if !providerIs("gce", "gke") {
-			By(fmt.Sprintf("Skipping reboot test, which is not implemented for %s", provider))
-			return
-		}
+	It("each node by ordering clean reboot and ensure they function upon restart", func() {
+		// clean shutdown and restart
+		testReboot(c, "sudo reboot")
+	})
 
-		// Get all nodes, and kick off the test on each.
-		nodelist, err := c.Nodes().List(labels.Everything(), fields.Everything())
-		if err != nil {
-			Failf("Error getting nodes: %v", err)
-		}
-		result := make(chan bool, len(nodelist.Items))
-		for _, n := range nodelist.Items {
-			go rebootNode(c, provider, n.ObjectMeta.Name, result)
-		}
+	It("each node by ordering unclean reboot and ensure they function upon restart", func() {
+		// unclean shutdown and restart
+		testReboot(c, "echo b | sudo tee /proc/sysrq-trigger")
+	})
 
-		// Wait for all to finish and check the final result.
-		failed := false
-		// TODO(mbforbes): Change to `for range` syntax and remove logging once
-		// we support only Go >= 1.4.
-		for _, n := range nodelist.Items {
-			if !<-result {
-				Failf("Node %s failed reboot test.", n.ObjectMeta.Name)
-				failed = true
-			}
-		}
-		if failed {
-			Failf("Test failed; at least one node failed to reboot in the time given.")
-		}
+	It("each node by triggering kernel panic and ensure they function upon restart", func() {
+		// kernel panic
+		testReboot(c, "echo c | sudo tee /proc/sysrq-trigger")
+	})
+
+	It("each node by switching off the network interface and ensure they function upon switch on", func() {
+		// switch the network interface off for a while to simulate a network outage
+		testReboot(c, "sudo ifdown eth0 && sleep 120 && sudo ifup eth0")
```

Comment:
I did run the test and all cases are passing. Probably the timeout for ssh session is longer then 2 minutes. Still, adding nohup will make it more robust.